modules/auxiliary/server/ldap_server.rb

```diff
@@ -0,0 +1,100 @@
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::LDAP::Server
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Native LDAP Server (Example)',
+        'Description' => %q{
+          This module provides a Rex based LDAP service to expose the
+          native Rex LDAP server functionality created during log4shell
+          development.
+        },
+        'Author' => [
+          'RageLtMan <rageltman[at]sempervictus>', # infrastructure
+          'Spencer McIntyre' # syntactically sane/correct Ruby LDAP object actions from early effort on l4j2 scanner
+        ],
+        'License' => MSF_LICENSE,
+        'References' => [],
+        'Actions' => [
+          [ 'Service', { 'Description' => 'Run LDAP server' } ]
+        ],
+        'PassiveActions' => [
+          'Service'
+        ],
+        'Notes' => {
+          'Stability' => [],
+          'Reliability' => [],
+          'SideEffects' => []
+        }
+      )
+    )
+  end
+
+  #
+  # Wrapper for service execution and cleanup
+  #
+  def run
+    start_service
+    service.wait
+  rescue Rex::BindFailed => e
+    print_error "Failed to bind to port #{datastore['SRVPORT']}: #{e.message}"
+  ensure
+    stop_service
+  end
+
+  #
+  # Creates Proc to handle incoming requests
+  #
+  def on_dispatch_request(client, data)
+    return if data.strip.empty?
+
+    data.extend(Net::BER::Extensions::String)
+    begin
+      pdu = Net::LDAP::PDU.new(data.read_ber!(Net::LDAP::AsnSyntax))
+      vprint_status("LDAP request data remaining: #{data}") if !data.empty?
+      resp = case pdu.app_tag
+             when Net::LDAP::PDU::BindRequest # bind request
+               # vprint_good("Received LDAP bind request from #{client} - #{pp pdu}")
+               client.authenticated = true
+               service.encode_ldap_response(
+                 pdu.message_id,
+                 Net::LDAP::ResultCodeSuccess,
+                 '',
+                 '',
+                 Net::LDAP::PDU::BindResult
+               )
+             when Net::LDAP::PDU::SearchRequest # search request
+               if client.authenticated || datastore['LDAP_AUTH_BYPASS']
+                 # Perform query against some loaded LDIF structure
+                 filter = Net::LDAP::Filter.parse_ldap_filter(pdu.search_parameters[:filter])
+                 attrs = pdu.search_parameters[:attributes].empty? ? :all : pdu.search_parameters[:attributes]
+                 res = service.search_ldif(filter, pdu.message_id, attrs)
+                 if res.nil? || res.empty?
+                   service.encode_ldap_response(pdu.message_id, Net::LDAP::ResultCodeNoSuchObject, '', 'No such object', Net::LDAP::PDU::SearchResult)
+                 else
+                   # Send the results and return success message for callback completion
+                   cli.write(res.join)
```

Comment:
So I have still have to update the msf.ldif file to change `objectClass` to `objectclass` in order to get results from this module, and afterwards it runs into this typo here where `cli` should be `client`.

```suggestion
                   client.write(res.join)
```

Reply:
did you restart framework? the ldif loader is in lib and taken from net-ldap's repo - they downcase in their code, but with the Java callbacks the casing is actually correct and the downcased keys don't match a thing on the filter :-\. How do you want to slice that? Should we enforce properly cased requests or downcase the keys in the search and on ingest?

Reply:
To elaborate a bit - Net::LDAP builds a Filter object which takes the requested attributes and then has a `match` method to find them in the ldif. If you downcase what goes into the ldif, then you wont match on camel-cased requests since those are what the filter is built around.

Reply:
the cli typo is fixed - thanks.

Reply:
Well [per the ldap wiki](https://ldapwiki.com/wiki/Distinguished%20Name%20Case%20Sensitivity) it sounds like it's up to us.

> Distinguished Name Case Sensitivity is dependent on the LDAP Server Implementation and the LDAP Protocol Exchange you are performing.

Knowing that, I'd vote to downcase on search and ingest.

Reply:
So that might leave us with a problem wherein `CN=firstwrite` will be overwritten by `cn=lastwrite` on ingest and the match on `CN` will always return the wrong `cn`

Reply:
Alright, so if we're keeping it as it is, we can validate this code path by changing the filter that `ldapsearch` uses by executing: `ldapsearch -x -H ldap://192.168.159.128 -b "dc=metasploit,dc=com" "(objectClass=*)"` which demonstrates that the module is working as intended.